### ISSUE TITLE
Vector sensor shape

### ISSUE THREAD

**Issue by CubeMD** [Unity-Technologies/ml-agents#4615]
**Is your feature request related to a problem? Please describe.**
Recently there has been a lot of great progress related to new types of sensors and perception systems for ML-Agents. Multidimensional perception (ie one used in grid sensor) seems very promising and has been successfully used for efficient feature encoding for complex data (ie skeleton). However to get the advantage of this new system one needs to implement their own version of the sensor which might be unnecessarily complex. Unless the environment is physics-based (which I should admit makes a big chunk of problems solved by the community) sensors are hard to reuse and require modifications for every new project. The majority of projects that require observations more complex than rays use good old Vector Sensor as a way to manage agent observations, therefore they will not benefit from this great new feature. 

**Describe the solution you'd like**
As far as I understand there is work in progress related to sensors and actuators and this might already be part of the roadmap, however, I would like to propose implementing the possibility to select dimensionality of the Vector Sensor in Behavior Parameters in addition or instead of Space Size setting.

**Describe alternatives you've considered**
Of course changing such a core system will increase the difficulty of the basic way of creating new observations and complicate the calculation of space size for agents. On top of that, a portion of precious tutorials describing initial steps with ML-Agents is going to become inaccurate. One alternative to changing Vector Sensor and Behavior Parameters could be the introduction of new documentation related to writing custom sensors, which is needed anyway. 

**Additional context**
A small video I would like to share, that describes nice results people have been able to achieve using network architecture described at 1:28. I believe sensors in the extensions folder could take advantage of a similar approach.
[https://www.youtube.com/watch?v=ym8Tnmiz5N8](url)

**Comment by CubeMD** [comment]
After some thoughts and discovering match3 branch, I decided to add a couple of things. I might have been too harsh with the wording of sensors being hard to reuse. What I meant is that having a board sensor is a good idea, as this functionality is common for many games, however, there might be "leftover" observations which might not fit in (eg turns left). I guess if a leftover observation is complex enough to require dimensionality it would deserve a sensor of its own anyway.
So far I have only used Vector Sensor for my projects, and I was afraid of missing out on a new feature. After taking a look at sensor implementation for match3 I think that writing own sensors might be a wiser decision.

**Comment by ervteng** [comment]
Hi @CubeMD, this is actually a great point and adding new sensor types is something we're actively working on. One of the tricks is that for each sensor type, the model side has to be able to create the appropriate encoder. With certain sensor dimensions, this is a real challenge. If you have a sensor that you've implemented, we'd be super happy to help merge it into the repo. 

If I may ask, what is your current use-case for the sensor you're looking to implement?

**Comment by CubeMD** [comment]
Hello @ervteng, currently I am making a game that is somewhat similar to chess, however, there is quite a bit more stuff on the board and there are effects that can be applied to the pieces. 

From the previous projects, I know that observation size and structure are playing a crucial role in training efficiency. In my case, there are ~15 attributes related to each piece (compared to ~5 per tile), so I decided to separate pieces observations from tiles as most of them are empty at any given time. With such an approach, I was able to reduce the observation space size from 300 to 125.

Long story short I was able to train a decent AI in ~12 hours with 8 total environments running in parallel. I might have achieved better results if I continued training, however, after a more careful inspection of AI's playstyle I concluded that my approach might have over complicated things or was not sufficient. Also at a time, there was a bug due to which I was unable to resume training anyway (thank you for the fix).

I came around the grid and match3 sensors conveniently right as I started thinking about using convolutions for solving my problem and currently testing this approach to see if it produces better results.

In your reply, you mentioned that an appropriate encoder needs to be created for each sensor type. I am interested in the internal workings of ML-Agents and studying both C# and Python sides of things. Could you point me to any place where it is done for any sensor? 

I would be extremely happy to contribute to ML-Agents. I have been working with it since 0.4 and have made several projects during the time.

**Comment by ervteng** [comment]
HI @CubeMD, that's great to hear! 12 hours is actually not bad for such a large observation space. Definitely give the Match 3 sensors a go - they've been merged into `master`, and should speed learning up quite a bit. 

Encoders are created in Python here:  https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/torch/utils.py#L144

The type of encoder is currently determined simply by its dimension - a 3D sensor is given a CNN, 1d is given a fully-connected network. You can imagine adding new types of encoders for 2D observations as well. We're looking to change this in the _very_ near future with more sophisticated type detection, so stay tuned!

**Comment by CubeMD** [comment]
@ervteng, thank you for encouragement and suggestion. I am working on adapting match3 sensor to my game right now. If it works out I would like to share it.
